lms/services/key.py

```diff
@@ -0,0 +1,63 @@
+import json
+
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from jose import constants, jwk
+from sqlalchemy import func
+
+from lms.models import Key
+from lms.services import AESService
+
+
+class KeyService:
+    def __init__(self, db, aes_service):
+        self._db = db
+        self._aes = aes_service
+
+    def generate(self) -> Key:
+        rsa_key = rsa.generate_private_key(
+            public_exponent=65537, key_size=2048, backend=default_backend()
+        )
+
+        # We are going to store the private key as an AES encoded PEM
+        pem_private_key = rsa_key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.TraditionalOpenSSL,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+        aes_iv = self._aes.build_iv()
+        aes_pem_private_key = self._aes.encrypt(aes_iv, pem_private_key)
+
+        # And the public key as a JWK json string
+        pem_public_key = rsa_key.public_key().public_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PublicFormat.SubjectPublicKeyInfo,
+        )
+        jwk_public_key = jwk.RSAKey(
+            algorithm=constants.Algorithms.RS256, rsa_key=pem_public_key.decode("utf-8")
+        ).to_dict()
+
+        key = Key(
+            _private_key=aes_pem_private_key,
+            public_key=json.dumps(jwk_public_key),
+            aes_cipher_iv=aes_iv,
+        )
+        self._db.add(key)
+        return key
```

Comment:
This is quite a lot in one go. It might help readability to split this into:

* `def _encrypt_private_key_as_pem(self, rsa_key, aes_iv)`
* `def _encrypt_public_key_as_jwk(self, rsa_key)`

Or something to break this up a bit